vm/src/obj/objbyteinner.rs

```diff
@@ -0,0 +1,349 @@
+                && self
+                    .elements
+                    .iter()
+                    .filter(|x| !char::from(**x).is_whitespace())
+                    .all(|x| char::from(*x).is_uppercase()),
+        ))
+    }
+
+    pub fn istitle(&self, vm: &VirtualMachine) -> PyResult {
+        if self.elements.is_empty() {
+            return Ok(vm.new_bool(false));
+        }
+
+        let mut iter = self.elements.iter().peekable();
+        let mut prev_cased = false;
+
+        while let Some(c) = iter.next() {
+            let current = char::from(*c);
+            let next = if let Some(k) = iter.peek() {
+                char::from(**k)
+            } else if current.is_uppercase() {
+                return Ok(vm.new_bool(!prev_cased));
+            } else {
+                return Ok(vm.new_bool(prev_cased));
+            };
+
+            let is_cased = current.to_uppercase().next().unwrap() != current
+                || current.to_lowercase().next().unwrap() != current;
+            if (is_cased && next.is_uppercase() && !prev_cased)
+                || (!is_cased && next.is_lowercase())
+            {
+                return Ok(vm.new_bool(false));
+            }
+
+            prev_cased = is_cased;
+        }
+
+        Ok(vm.new_bool(true))
+    }
+
+    pub fn lower(&self, _vm: &VirtualMachine) -> Vec<u8> {
+        self.elements.to_ascii_lowercase()
+    }
+
+    pub fn upper(&self, _vm: &VirtualMachine) -> Vec<u8> {
+        self.elements.to_ascii_uppercase()
+    }
+
+    pub fn hex(&self, vm: &VirtualMachine) -> PyResult {
+        let bla = self
+            .elements
+            .iter()
+            .map(|x| format!("{:02x}", x))
+            .collect::<String>();
+        Ok(vm.ctx.new_str(bla))
+    }
+
+    pub fn fromhex(string: String, vm: &VirtualMachine) -> Result<Vec<u8>, PyObjectRef> {
+        // first check for invalid character
+        for (i, c) in string.char_indices() {
+            if !c.is_digit(16) && !c.is_whitespace() {
+                return Err(vm.new_value_error(format!(
+                    "non-hexadecimal number found in fromhex() arg at position {}",
+                    i
+                )));
+            }
+        }
+
+        // strip white spaces
+        let stripped = string.split_whitespace().collect::<String>();
+
+        // Hex is evaluated on 2 digits
+        if stripped.len() % 2 != 0 {
+            return Err(vm.new_value_error(format!(
+                "non-hexadecimal number found in fromhex() arg at position {}",
+                stripped.len() - 1
+            )));
+        }
+
+        // parse even string
+        Ok(stripped
+            .chars()
+            .collect::<Vec<char>>()
+            .chunks(2)
+            .map(|x| x.to_vec().iter().collect::<String>())
+            .map(|x| u8::from_str_radix(&x, 16))
+            .map(|x| x.unwrap())
+            .collect::<Vec<u8>>())
+    }
+}
+
+// TODO
+// fix b"é" not allowed should be bytes("é", "utf8")
```

Comment:
it's not allowed anymore, right? Is this TODO still relevant?

Reply:
no it's ok now